pkg/jobs/job.go

```
@@ -0,0 +1,307 @@
+package jobs
+
+import (
+	"context"
+	"io"
+	"os"
+	"os/exec"
+	"syscall"
+	"time"
+)
+
+type jobState struct {
+	job    Job
+	result JobResult
+	cmd    *exec.Cmd
+	b      *Backend
+}
+
+type Job struct {
+	Id        JobId
+	Uname     UserName
+	Command   Command
+	StartTime int64
+	Pid       int
+}
+
+type JobResult struct {
+	EndTime int64
+	Status  Status
+	Code    int
+}
+
+func newRunningJobState(b *Backend, j *Job, c *exec.Cmd) *jobState {
+	return &jobState{
+		job:    *j,
+		result: JobResult{EndTime: -1, Status: Running, Code: -1},
+		cmd:    c,
+		b:      b,
+	}
+}
+
+func startNewJob(ctx context.Context, b *Backend, u *UserName, c *Command) (*JobId, error) {
+	id, err := NewJobId()
+	if err != nil {
+		return nil, err
+	}
+
+	j := &Job{Id: *id, Uname: *u, Command: *c, StartTime: -1, Pid: -1}
+
+	_, err = exec.LookPath(c.Cmd)
+	if err != nil {
+		_ = j.writeErrorStatus(ctx, b, NotFound, -1)
+		return id, nil
+	}
+
+	cmd := exec.Command(c.Cmd, c.Args...)
+
+	out, err := b.jobLogs.create(ctx, u, id)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd.Stdout = out
+	cmd.Stderr = out
+	cmd.Dir = c.Dir
+	// Not supported at the moment authorize and set credentials
+	// SysProcAttr.Credential
+	// do not kill child if server is stopped
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true, Pgid: 0}
+
+	err = cmd.Start()
+	if err != nil {
+		_ = j.writeErrorStatus(ctx, b, NotFound, -1)
+		return id, nil
+	}
+
+	startTime := time.Now().Unix()
+
+	b.wg.Add(1)
+
+	j.StartTime = startTime
+	j.Pid = cmd.Process.Pid
+
+	err = b.repo.CreateJob(ctx, j)
+	if err != nil {
+		log.Error("Can't write a running job", err)
+		return nil, err
```

Error handling should be consistent. I noticed above you do `j.writeErrorStatus` and return `nil`, but here for an error but here you return the error.

Ideally you would wrap this function in another function. Then you can just return an error here and if the error is not `nil` then return the appropriate GRPC response.

---

The idea here that the entire jobs package has no idea about GRPC. It operates on his own set of structures/errors. API package does have a clear understanding about that and map jobs package structures/errors to the corresponding GRPC stuff.